File: programmers/lv3/양과 늑대.py
from collections import defaultdict

def solution(info, edges):
    answer, node_dict = 0, defaultdict(list)
    
    for edge in edges:
        node_dict[edge[0]].append(edge[1])
    
    def dfs(wolves, sheeps, cand_set):
        nonlocal answer
        check = False
        for i in cand_set:
            if sheeps - wolves - info[i] > 0:
                new_cand_list = list(cand_set)
                new_cand_list.remove(i)
                new_cand_list.extend(node_dict[i])
                new_cand_set = set(new_cand_list)
                check = True
                if info[i] == 0:
                    dfs(wolves, sheeps + 1, new_cand_set)
                else:
                    dfs(wolves + 1, sheeps, new_cand_set)
                
        if not check:
            answer = max(answer, sheeps)
    
    dfs(0, 1, set(node_dict[0]))
    
    return answer

File: programmers/lv3/양과 늑대.java
import java.util.*;

class Solution {
    int maxSheep = 1;
    Map<Integer, List<Integer>> children = new HashMap<>();
    
    void dfs(int[] info, Set<Integer> includeSet, int sheeps, int wolves) {
        Set<Integer> newSet = new HashSet<>(includeSet);
        
        for(int idx: includeSet) {
            newSet.add(idx);
            for(int child: children.get(idx)) {
                if(!includeSet.contains(child)) {
                    if(info[child] == 0 || (info[child] == 1 && sheeps - wolves > 1)) {
                        newSet.add(child);
                        int newSheeps = sheeps, newWolves = wolves;
                        if(info[child] == 0) {
                            newSheeps += 1;
                            maxSheep = Math.max(maxSheep, newSheeps);
                        } else {
                            newWolves += 1;
                        }
                        dfs(info, newSet, newSheeps, newWolves);
                        newSet.remove(child);
                    }
                }
            }
        }
    }
    
    public int solution(int[] info, int[][] edges) {
        int answer = 0;
        for(int i = 0; i < info.length; i++) {
            children.put(i, new ArrayList<>());
        }
        
        for(int[] edge: edges) {
            children.get(edge[0]).add(edge[1]);
        }
        
        Set<Integer> set = new HashSet<>();
        set.add(0);
        
        dfs(info, set, 1, 0);
        
        answer = maxSheep;
        
        return answer;
    }
}
